File: string_array/Rotate_Array.py
class Solution:
    def reverse(self, nums: List[int], j: int, k: int) -> None:
        while j < k:
            nums[j], nums[k] = nums[k], nums[j]
            j += 1
            k -= 1
    def rotate(self, nums: List[int], k: int) -> None:
        k = k % len(nums)
        if (k == 0):
            return
        self.reverse(nums, 0, len(nums) - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, len(nums) - 1)

File: string_array/Rotate_Array.java
class Solution {
    public void rotate(int[] nums, int k) {
        k = k % nums.length;

        if (k == 0) return;
        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, nums.length - 1);
    }

    private void reverse(int[] nums, int j, int k) {
        int swap;
        while (j < k) {
            swap = nums[k];
            nums[k--] = nums[j];
            nums[j++] = swap;
        }
    }
}